pkg/ingester/instance.go

```diff
@@ -690,6 +681,63 @@ func (i *instance) GetSeriesVolume(ctx context.Context, req *logproto.VolumeRequ
 	return res, nil
 }
 
+func prepareLabelsAndMatchers(targetLabels []string, matchers []*labels.Matcher) (map[string]struct{}, []*labels.Matcher, bool) {
+	if len(targetLabels) > 0 {
+		return prepareLabelsAndMatchersWithTargets(targetLabels, matchers)
+	}
+
+	includeAll := len(matchers) == 0
+	labelsToMatch := make(map[string]struct{})
+
+	for _, m := range matchers {
+		if m.Name == "" {
+			includeAll = true
+			continue
+		}
+
+		labelsToMatch[m.Name] = struct{}{}
+	}
+
+	return labelsToMatch, matchers, includeAll
+}
+
+func prepareLabelsAndMatchersWithTargets(targetLabels []string, matchers []*labels.Matcher) (map[string]struct{}, []*labels.Matcher, bool) {
+	matchAllIndex := -1
+	labelsToMatch := make(map[string]struct{})
+	targetsFound := make(map[string]bool, len(targetLabels))
+
+	for _, target := range targetLabels {
+		labelsToMatch[target] = struct{}{}
+		targetsFound[target] = false
+	}
+
+	for i, m := range matchers {
+		if m.Name == "" {
+			matchAllIndex = i
+			continue
+		}
+
+		if found, ok := targetsFound[m.Name]; ok && !found {
```

Comment:
nit:  We do a map access each time here anyway so it's probably faster to just set `targetsFound[m.Name]` each time. It would certainly be easier for me to read :slightly_smiling_face: